### ISSUE TITLE
Cypress sometimes stalls/hangs with no output when running in Jenkins with Docker

### ISSUE THREAD

**Issue by kaiyoma** [cypress-io/cypress#8206]
### Current behavior:

I feel like this kind of issue has been referenced many times before, but even after trying all the solutions I've found, we still continue to run into this problem. We've been seeing this issue for many months now, with early versions of Cypress 4.x and even with the most recent version today.

Long-running (multi-hour) test suites often "stall": the output stops appearing, but nothing actually fails or crashes. We actually have to rely on the timeout mechanism in Jenkins to kill the task because Cypress is just stuck.

I have enabled debug logs and honestly I don't see anything helpful in them. I'll upload the relevant portion next, but there are no mentions of failures, crashes, or even being out of memory. We're already following all the guidance in the Cypress docs about disabling /dev/shm and using the correct IPC setting.

### Desired behavior:

Tests should either run to completion or Cypress should fail with a clear error.

### Versions

Cypress: 4.12.1
OS: Linux
Browser: Chrome 83 (headless)

**Comment by kaiyoma** [comment]
Here's an abridged version of the full Cypress debug log: [cypress-debug-abridged.log](https://github.com/cypress-io/cypress/files/5036280/cypress-debug-abridged.log)

In our Jenkins output, this is the last message I see:

```
06:43:15        ✓ should render a read & write view for Image Management (3909ms)
```

You can see this mocha test at the top of the logfile. But after that, no more mocha tests are triggered for some reason. Nothing runs out of memory, nothing crashes, nothing fails, nothing exits. Everything just stops responding and Cypress seems to be waiting for an event that'll never happen.

I can furnish the full debug logfile (13.2 MB) upon request.

**Comment by AnkitPhalnikerTR** [comment]
We are facing similar issue, Cypress just hangs indefinitely during UI test runs, no error, no logs. The only option is to close and run the tests again. We are running them locally, have not integrated with CI yet.

**Comment by cellog** [comment]
Also seeing this, about 1 every 50 test runs, Cypress will hang, occasionally for 12 hours(!) if no one notices

**Comment by jennifer-shehane** [comment]
The debug logs don't indicate anything that we can go off of. We'll need a way to reproduce this in order to investigate.

**Comment by kaiyoma** [comment]
> The debug logs don't indicate anything that we can go off of. We'll need a way to reproduce this in order to investigate.

Run multi-hour test suites in a Jenkins/Docker environment until you start to see Cypress hanging. 😄

On a serious note, it sounds like several people are seeing this problem, and we all have access to such environments, so maybe you'll have to just leverage that rather than trying to repro yourself. Is there any other logging or debugging instrumentation we should turn on that would be helpful?

**Comment by anaisamp** [comment]
I've also experienced this.
I don't have a long-running/multi-hour test suite though, I only have 2 test files so far.
The tests have always passed locally. But when running inside a docker container, tests hanged most of the times, without any feedback, and I had to manually stop the Jenkins pipeline.

After a few days of trying things out, making these 3 changes helped me.

1. Not using Chrome. Before I was using `--browser chrome`. When I removed it, I saw a helpful output, at the same stage where the tests used to hang. 

<img width="966" alt="Screenshot 2020-08-18 at 19 47 35" src="https://user-images.githubusercontent.com/1133608/90553192-df394e80-e18b-11ea-90a4-a4d4eefa1d09.png">

2. I could see the error was related to a memory leak. So I added `ipc` and `shm_size` to the docker-compose file, as everyone was advising. (Had done it before, but seemed to have worked only after not using Chrome.)

```
version: '3.5'
services:
  app-build:
    ipc: host
    shm_size: '2gb'
```

3. I'm using fixtures, and I trimmed them to use smaller .json files.

Please note I'm just sharing what seemed to have worked for me, but I don't consider this a solution. It sounds like an issue that should be looked into.

**Comment by kaiyoma** [comment]
Interesting. To the best of my knowledge, we've seen the opposite trend: using Electron was worse and switching to Chrome has slightly improved things. But we still run into this hanging problem with both browsers. (This would make sense if the problem is actually with mocha, which is kind of what the logs would suggest.)

**Comment by bradyolson** [comment]
FWIW, we were seeing this nearly every test suite execution (>90%) in our CI runner (AWS CodeBuild with parallelization, using Chrome), with no issues running locally. Today we upgraded to v5.0.0 and the hanging/stalling has seemingly stopped

**Comment by kaiyoma** [comment]
@bradyolson Good to know! We're attempting to upgrade Cypress this week, so maybe in a few days I can report back with our results.

**Comment by cellog** [comment]
I have an update on this:

I noticed that Cypress tests were downloading a few static assets in a continuous stream. The site logo, a font, etc. They would be downloaded hundreds of times in a single test, and in some cases completely overwhelm the browser, causing the test to fail. This happened locally too.

I downgraded to version 4.9.0 from 4.12.0 and the issue went away. I hope this gives some context to help.

For some other context: this happened in end-to-end mode and in mock mode (we run our tests in both modes) AND in a new mode using mock service worker, but NOT when running the app in development or in production through a regular Chrome browser.

**Comment by kaiyoma** [comment]
We finally got Cypress 5 to run and hit the same exact problem. No improvement.

**Comment by kaiyoma** [comment]
I spent the afternoon trying to run our tests with Firefox (which I realize is in "beta") and noticed that this problem seems a lot worse with Firefox. It's the same symptoms as before: console output stops appearing, mocha events appear to stop coming in, but the debug log keeps chugging along with memory and CPU updates.

When trying to run our usual battery of ~50 test suites, we couldn't get further than 5 suites, sometimes failing at even the first one. With Chrome, we could always get at least 25 or 30 suites in before the stall happens.

**Comment by kaiyoma** [comment]
I've tried a new approach/workaround in our project where instead of passing all our test suites to Cypress all at once, I pass the test suites one at a time.

Before:

```
cypress run --browser chrome --headless --spec foo.spec.js,bar.spec.js,baz.spec.js
```

After:

```
cypress run --browser chrome --headless --spec foo.spec.js
cypress run --browser chrome --headless --spec bar.spec.js
cypress run --browser chrome --headless --spec baz.spec.js
```

This seems to have helped a little bit, though last night we hit an error even with this approach. About 2 hours into the testing, a test suite stalled after running for only a couple minutes and Jenkins had to kill the task a couple hours later.

**Comment by kaiyoma** [comment]
We encountered this failure again last night with the same symptoms. Now that we're running our test suites one at a time, the logs are smaller and easier to digest. Still seems like Cypress is at fault here.

In the Jenkins output, we can see that the test suite stalled after executing for only a minute:

```
03:20:20    (Run Starting)
03:20:20  
03:20:20    ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
03:20:20    │ Cypress:    5.0.0                                                                              │
03:20:20    │ Browser:    Chrome 84 (headless)                                                               │
03:20:20    │ Specs:      1 found (topology/overlay.spec.js)                                                 │
03:20:20    │ Searched:   cypress/integration/topology/overlay.spec.js                                       │
03:20:20    └────────────────────────────────────────────────────────────────────────────────────────────────┘
03:20:20  
03:20:20  
03:20:20  ────────────────────────────────────────────────────────────────────────────────────────────────────
03:20:20                                                                                                      
03:20:20    Running:  topology/overlay.spec.js                                                        (1 of 1)
03:20:20    Adding --disable-dev-shm-usage...
03:20:20    Adding --disable-gpu...
03:20:20    Adding --window-size=1920,1080...
03:20:38  
03:20:38  
03:20:38    Topology overlay tests
03:21:24      ✓ should navigate to Topology application (40059ms)
03:21:24      ✓ should have the right options in the select (352ms)
04:01:08  Cancelling nested steps due to timeout
04:01:08  Sending interrupt signal to process
04:01:20  Terminated
```

In the full debug logfile, I can see that this is the last event we get from mocha:

```
2020-09-03T10:21:18.281Z cypress:server:reporter got mocha event 'hook end' with args: [ { id: 'r5', title: '"before each" hook', hookName: 'before each', hookId: 'h2', body: ...
```

There are no more mocha events and the test never starts execution. This is also verified by the video file that Cypress writes out. I can see in the video that the sidebar never "expands" for the doomed test. Instead of seeing the list of `get` and `assert` calls, the spinner next to the test simply spins forever.

I can furnish the Jenkins log, Cypress debug log, and Cypress video upon request (in a DM or some other private channel).

**Comment by mayrbenjamin92** [comment]
We have the same issue and we are using Cypress 5.2.0 as it seems to bring in several performance fixes. We are now using Cypress since the 2.X days and I feel that this is getting worse over time.

**Comment by bahunov** [comment]
Any update on this issue?

**Comment by kaiyoma** [comment]
> Any update on this issue?

+1. Also waiting for updates. We continue to run into this issue often, even with the latest version (5.2.0). I've offered to furnish full logs, but no one's taken me up on it. Kind of feels like the Cypress folks are hoping this issue will just magically disappear.

**Comment by danielcaballero** [comment]
We are facing the same issue. Tried to update to 5.2.0 and all kind of workarounds but this is still happening.

**Comment by muratkeremozcan** [comment]
Similar issues here. There has been a runner change as well as upgrade to Cypress 5.2.0 . Firefox tests hang indefinitely. With Electron, we can get the same crashes, but rarely the tests can work. We have parallelization as well.

The trends seems to be tests with ui-login. For the time being we disabled them in CI. 

This might help someone.

```
const ignoreOnEnvironment = (environment: 'linux' | 'darwin' | 'win32', fn) => {
  if (Cypress.platform !== environment) {
    fn();
  }
};
```

Then, in the test wrap any describe/context/it block

```
ignoreOnEnvironment('linux', () => {
  describe(....)
});
```

**Comment by dragon788** [comment]
If you are running Chrome or Chromium it can be more useful to use `--disable-shm-usage` as a flag rather than trying to set the `--shm-size` for Docker or Docker-Compose, especially when running in Docker-in-Docker on Jenkins or other CI providers.

**Comment by muratkeremozcan** [comment]
> If you are running Chrome or Chromium it can be more useful to use `--disable-dev-shm-usage` as a flag rather than trying to set the `--shm-size` for Docker or Docker-Compose, especially when running in Docker-in-Docker on Jenkins or other CI providers.
> 
> https://developers.google.com/web/tools/puppeteer/troubleshooting#tips
> 
> [#5336 (comment)](https://github.com/cypress-io/cypress/issues/5336#issue-505290346)

This helped. Thank you!

The infor resource can also be found that at Cypress [docs](https://docs.cypress.io/guides/guides/continuous-integration.html#In-Docker). If your tests are hanging in the CI, take a look.

There are plenty of questions about configuring the `plugins/index.js` file. Here is a complex example, not even using the new `--config-file` pattern, instead using `getConfigurationByFile` pattern.

```javascript
const fs = require('fs-extra')
const path = require('path')
// tasks go here
const percyHealthCheck = require('@percy/cypress/task')
const mailosaurTasks = require('./mailosaur-tasks')
// merge all tasks
const all = Object.assign({},
  percyHealthCheck,
  mailosaurTasks
)

function getConfigurationByFile(file) {
  const pathToConfigFile = path.resolve('cypress/config', `${file}.json`)

  return fs.readJson(pathToConfigFile)
}

module.exports = (on, config) => {
  on("task", all);
  // needed to address issues related to tests hanging in the CI
  on('before:browser:launch', (browser = {}, launchOptions) => {
    if (browser.family === 'chromium' && browser.name !== 'electron') {
      launchOptions.args.push('--disable-dev-shm-usage')
    }
    return launchOptions
  });

  const file = config.env.configFile || 'dev'
  return getConfigurationByFile(file)
}
```

**Comment by AlexanderTunick** [comment]
Similar issue here. Already throughout months.
Cypress version `"cypress": "5.1.0"`

From time to time (several times in a week) the execution just stops on random tests.
![Screenshot 2020-12-04 at 11 10 38](https://user-images.githubusercontent.com/29304271/101144700-64ece700-3621-11eb-9ff4-4643d4640bfd.png)

No errors, no memory ran out. Just stopped and JENKINS continues its execution for "dead tests".

We checked a Server, Docker, Jenkins, and nothing outstanding there.

**Comment by Garreat** [comment]
#https://github.com/cypress-io/cypress/issues/9350#issuecomment-739045845

**Comment by prashantabellad** [comment]
@bahmutov @jennifer-shehane Can anyone from Cypress team update on this?

**Comment by gururajhm** [comment]
Guys me to facing same issue. I have been using cypress for almost 1 yr 4 months never faced this issue, I updated to latest version then it keeps hanging or some time memory issues in local and in Circle CI too. Badly need a fix...

-G

**Comment by hermape7** [comment]
I have spotted this also in the CircleCI jobs. 
I have disabled the shm-usage for the chrome (`launchOptions.args.push("--disable-dev-shm-usage");`)
This "tests hanging" is occurring from time to time only (1 from 30 runs).

**Comment by alexf101** [comment]
We're also experiencing this issue - about two to three minutes into a test suite that's otherwise working fine, the test runner hangs indefinitely. With debug logs enabled, there's absolutely nothing of interest to see, all logs stop except process_profiler, and there's no error message. I can reproduce this 100% of the time, the only variance is at what point in the test suite it'll hang.

The test suite in question uses the visual diff plugin to take snapshots of some of our UIKit reusable components in various configurations.

When I play with the shm_size or --disable-dev-shm-usage flag it doesn't have much affect - maybe got through a little further on average using a large shm as opposed to disabling shm.

Cypress 5 or 6 didn't make a difference either, both have this symptom.

What *did* help was making the test suite smaller - I split it three ways so that each suite took under a minute, and now it passes reliably.

This is a really unstable solution, as some unsuspecting developer is going to come along and add some more tests to an existing suite, and the result is going to be that the test suite in question starts hanging indefinitely and ruin their day.

I'm happy to provide DEBUG logs or anything like that if it helps find a solution - as I said, I can reproduce this hang reliably. I can't send you my Docker container though as our application code under test is proprietary.

**Comment by prashantabellad** [comment]
@bahmutov @jennifer-shehane Please can this be prioritized else this is a dampener

**Comment by tsuna** [comment]
> What _did_ help was making the test suite smaller - I split it three ways so that each suite took under a minute, and now it passes reliably.

Anecdotally, this is also our workaround. We used to run all our test suites in a single go, and this would cause Cypress to lock up much more often. Now we run each test suite separately, and the problem still occurs but a lot less frequently. It's very annoying because there is a huge startup cost to spin up Cypress for each test suite, so this makes our entire end-to-end tests take 3h instead of 2h or less.

We also disable pass `--disable-dev-shm-usage` as Docker containers typically have a tiny `/dev/shm`, and we run everything in them.

**Comment by bahunov** [comment]
> > What _did_ help was making the test suite smaller - I split it three ways so that each suite took under a minute, and now it passes reliably.
> 
> Anecdotally, this is also our workaround. We used to run all our test suites in a single go, and this would cause Cypress to lock up much more often. Now we run each test suite separately, and the problem still occurs but a lot less frequently. It's very annoying because there is a huge startup cost to spin up Cypress for each test suite, so this makes our entire end-to-end tests take 3h instead of 2h or less.
> 
> We also disable pass `--disable-dev-shm-usage` as Docker containers typically have a tiny `/dev/shm`, and we run everything in them.

Have you guys tried with garbage collection in chrome, also set noTestKeepInMem to 0

**Comment by dragon788** [comment]
One thing that we ran into was the previously silent failures of doing things in beforeEvent or afterEvent hooks would also cause this type of exception/hang.

There were some warnings and some handling added is exceptions occur outside the normal flow, but a lot of times the issue happens if you are trying to look at some state in a spot where it might not be ready or has already been discarded.

**Comment by ankit-madan** [comment]
> > > What _did_ help was making the test suite smaller - I split it three ways so that each suite took under a minute, and now it passes reliably.
> > 
> > 
> > Anecdotally, this is also our workaround. We used to run all our test suites in a single go, and this would cause Cypress to lock up much more often. Now we run each test suite separately, and the problem still occurs but a lot less frequently. It's very annoying because there is a huge startup cost to spin up Cypress for each test suite, so this makes our entire end-to-end tests take 3h instead of 2h or less.
> > We also disable pass `--disable-dev-shm-usage` as Docker containers typically have a tiny `/dev/shm`, and we run everything in them.
> 
> My issue disappeared when I did following, (still using v.5 with server/route feature):
> 
> 1.Run garbage collection after each - works in chrome not sure about electron.
> 2. Set numTestsKeptInMemory to 0

How do you run garbage collection in Cypress?

**Comment by kaiyoma** [comment]
It's been a few months since I commented, so I'll add an update. We still see this problem decently often: not a ton, but enough to possibly consider Cypress alternatives. We see this problem at least once a day and "Just retrigger the task" is becoming a tiresome refrain.

We've been able to track down a couple of our recurring stalls to our AUT itself. In some cases, our UI is doing a lot of React re-rendering, data crunching, or other high-CPU activities, and the browser seems to be hogging all the CPU and starving the Cypress process. If others are seeing these stalls happening in the same tests over and over, I'd suggest looking into a high CPU problem.

However, the majority of these hangs happen in a different test every time, usually when testing a part of the UI that isn't particularly resource-intensive. I'm looking at an example from 30 minutes ago and I can see that the test stalled 90 seconds into execution, which probably rules out any kind of memory leak or OOM situation.

Any advice from the Cypress folks about things we could do on our end to gather more information for you guys?

**Comment by prashantabellad** [comment]
@jennifer-shehane @bahmutov any updates/comments from your side?

**Comment by retypepassword** [comment]
> > My issue disappeared when I did following, (still using v.5 with server/route feature):
> > 1.Run garbage collection after each - works in chrome not sure about electron.
> > 2. Set numTestsKeptInMemory to 0
> 
> How do you run garbage collection in Cypress?

See https://github.com/cypress-io/cypress/issues/8525

**Comment by retypepassword** [comment]
We've been experiencing this issue in CircleCI (stalls/hangs with no output) with Cypress v6 and Chrome. I have logs from `socket.io:socket` to add (see below) if it helps.

I've noticed this sequence:
1. ["socket connected" event](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L201) fires (let's call this _socket 1_)
2. `automation:client:connected` fires and [`automationClient = socket` occurs](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L213).
3. "socket connected" fires again (_socket 2_)
4. `automationClient.on('disconnect', …)` on [this line](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L219) doesn't fire when _socket 2_ disconnects and no error message is logged. All the other events in the `Socket` class are handled using _socket 2_ (because they use `socket.on(),` not `automationClient.on()`)

For what it's worth, adding `socket.on('disconnect', …)` to the top level of the `this.io.on('connection', (socket) => {` block catches the disconnect event. I've also seen Cypress hang with no output if the socket disconnects and then reconnects.

```log
────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  [redacted]_spec.ts                                                         (5 of 13)
socket connected - writing packet <------- socket connected
joining room uH9QsXHm0ty8jFKzAAAI
packet already sent in initial handshake
joined room [ 'uH9QsXHm0ty8jFKzAAAI' ]
got packet {"type":2,"nsp":"/","data":["automation:client:connected"]} <-------- automationClient = socket
emitting event ["automation:client:connected"]
dispatching an event ["automation:client:connected"]
socket connected - writing packet <-------- socket connected
joining room pCQJu9N7VhNgLysHAAAJ
packet already sent in initial handshake
joined room [ 'pCQJu9N7VhNgLysHAAAJ' ]
got packet {"type":2,"nsp":"/","data":["runner:connected"]}
emitting event ["runner:connected"]
dispatching an event ["runner:connected"]
joining room runner
joined room [ 'runner' ]
got packet {"type":2,"nsp":"/","id":0,"data":["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"}]}
emitting event ["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"}]
attaching ack callback to event
dispatching an event ["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"},null]
sending ack [true]
got packet {"type":2,"nsp":"/","data":["app:connect","wfk9l"]}
emitting event ["app:connect","wfk9l"]
dispatching an event ["app:connect","wfk9l"]
got packet {"type":2,"nsp":"/","data":["spec:changed","integration/fosLoadDetails_spec.ts"]}
emitting event ["spec:changed","integration/[redacted]_spec.ts"]
dispatching an event ["spec:changed","integration/[redacted]_spec.ts"]
got packet {"type":2,"nsp":"/","data":["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]}
emitting event ["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]
dispatching an event ["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]
closing socket - reason transport close
```

**Comment by retypepassword** [comment]
We've been experiencing this issue in CircleCI (stalls/hangs with no output) with Cypress v6 and Chrome. I have logs from `socket.io:socket` to add (see below) if it helps.

I've noticed this sequence:
1. ["socket connected" event](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L201) fires (let's call this _socket 1_)
2. `automation:client:connected` fires and [`automationClient = socket` occurs](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L213).
3. "socket connected" fires again (_socket 2_)
4. `automationClient.on('disconnect', …)` on [this line](https://github.com/cypress-io/cypress/blob/c37896089b23bd083c9fae5770ea638f87b0deb8/packages/server/lib/socket.js#L219) doesn't fire when _socket 2_ disconnects and no error message is logged. All the other events in the `Socket` class are handled using _socket 2_ (because they use `socket.on(),` not `automationClient.on()`)

For what it's worth, adding `socket.on('disconnect', …)` to the top level of the `this.io.on('connection', (socket) => {` block catches the disconnect event. I've also seen Cypress hang with no output if the socket disconnects and then reconnects.

```log
────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  [redacted]_spec.ts                                                         (5 of 13)
socket connected - writing packet <------- socket connected
joining room uH9QsXHm0ty8jFKzAAAI
packet already sent in initial handshake
joined room [ 'uH9QsXHm0ty8jFKzAAAI' ]
got packet {"type":2,"nsp":"/","data":["automation:client:connected"]} <-------- automationClient = socket
emitting event ["automation:client:connected"]
dispatching an event ["automation:client:connected"]
socket connected - writing packet <-------- socket connected
joining room pCQJu9N7VhNgLysHAAAJ
packet already sent in initial handshake
joined room [ 'pCQJu9N7VhNgLysHAAAJ' ]
got packet {"type":2,"nsp":"/","data":["runner:connected"]}
emitting event ["runner:connected"]
dispatching an event ["runner:connected"]
joining room runner
joined room [ 'runner' ]
got packet {"type":2,"nsp":"/","id":0,"data":["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"}]}
emitting event ["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"}]
attaching ack callback to event
dispatching an event ["is:automation:client:connected",{"element":"__cypress-string","string":"0.25538962826612965"},null]
sending ack [true]
got packet {"type":2,"nsp":"/","data":["app:connect","wfk9l"]}
emitting event ["app:connect","wfk9l"]
dispatching an event ["app:connect","wfk9l"]
got packet {"type":2,"nsp":"/","data":["spec:changed","integration/[redacted]_spec.ts"]}
emitting event ["spec:changed","integration/[redacted]_spec.ts"]
dispatching an event ["spec:changed","integration/[redacted]_spec.ts"]
got packet {"type":2,"nsp":"/","data":["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]}
emitting event ["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]
dispatching an event ["watch:test:file",{"name":"[redacted]_spec.ts","relative":"cypress/integration/[redacted]_spec.ts","absolute":"/home/********/********/cypress/integration/[redacted]_spec.ts","specType":"integration"}]
closing socket - reason transport close
```

**Comment by norvinino** [comment]
@jennifer-shehane Cypress closes automatically and stops working when a test is run

**Comment by pnilesh10** [comment]
We moved to V 5.3.0 and started using Electron instead of chrome.
Didn't encounter any stalls since then.

**Comment by KeKs0r** [comment]
I have tried running my e2e tests on 3x bigger containers, which did not seem to have an impact as well as running garbage collection between tests. So I am not sure if this is only caused by ressource limitation or there are other things at play.

**Comment by heikomat** [comment]
We were noticing the same symptom in our tests. When running on CI, the runners often just stop working after a while.

In our case the problem was **not with cypress, but with the host machine**

Looking at `/varlog/syslog` and `dmesg` we noticed a lot of logs that logs that looked like this:
```
[11512495.394975] TCP: out of memory -- consider tuning tcp_mem
[11512581.913531] TCP: out of memory -- consider tuning tcp_mem
[11512616.330274] TCP: out of memory -- consider tuning tcp_mem
[11512640.837935] TCP: out of memory -- consider tuning tcp_mem
[11512676.482550] TCP: out of memory -- consider tuning tcp_mem
[11512748.301204] TCP: out of memory -- consider tuning tcp_mem
```

After reading [this blogpost](https://dzone.com/articles/tcp-out-of-memory-consider-tuning-tcp-mem) we looked for more info on the `rmem_max` and `wmem_max` values and decided to pick the values from [this article](https://www.tecchannel.de/a/tcp-ip-tuning-fuer-linux,429773,6).

It basically explains these values and ends with
> To change them, put this into your `/etc/sysctl.conf`
> ```
> # increase Linux TCP buffer limits
> net.core.rmem_max = 2097152
> net.core.wmem_max = 2097152
> 
> # increase Linux autotuning TCP buffer limits
> # min, default, and max number of bytes to use
> net.ipv4.tcp_rmem = 4096 87380 2097152
> net.ipv4.tcp_wmem = 4096 65536 2097152
> ```
> And run `sysctl -p` to apply them


So we did, and since then two things changed
1. The `TCP: out of memory`-logs completely stopped appearing
2. The cypress runners no longer stopped working mid-run and our tests went smooth

**Comment by bahunov** [comment]
I had similar issues in the pipeline but it turned out to be the Java
backend that was eating up all of the resources.
The problem got fixed when I put limit on resource for the backend service.

**Comment by kaiyoma** [comment]
Since I posted this originally, I'll post an update of where we're at with this issue. We're on Cypress 9 now, and we're also using the Cypress Docker image (which was not true when I first posted this). Things have definitely gotten better since the summer of 2020 and now our Cypress tests are actually quite reliable.

We do run into this stalling issue occasionally, but now when it happens, it always points to a problem with the application under test, and it's always a high CPU issue. (For a long time, we thought it could be memory-related, and tried all sorts of things to address that, but they never made a difference.) Whenever we start seeing Cypress stall for a test, we'll investigate that specific part of the UI, and we always find that it's pretty sluggish, either because it's stuck in a rendering loop or doing too much other processing.

I'll note that we're still running all our Cypress tests one at a time (invoking `cypress` separately for each one). I haven't tried going back to running all of them at once since we're pretty stable at the moment and I don't want to rock the boat. For now, we're pretty content since we know what the stalling issue means for us. I wish the error/failure was more direct and obvious, but we have enough expertise with Cypress now to recognize this situation.

**Comment by tsuna** [comment]
On Fri, Feb 4, 2022 at 5:05 PM Kyle Getz ***@***.***> wrote:

> I'll note that we're still running all our Cypress tests one at a time
> (invoking cypress separately for each one). I haven't tried going back to
> running all of them at once since we're pretty stable at the moment and I
> don't want to rock the boat
>
Well, if things are working reliably now, maybe it's time to try to run
cypress just once for all the tests, as this would save a lot of time. And
if we reproduce the issue in doing so, we can always roll back and report
back here, as it would still point to an issue inherent to Cypress.

**Comment by martincarstens** [comment]
We were having the same issue in GitHub Actions. I wanted to post this here in case it helps anyone. I haven't seen this approach mentioned so wanted to make sure it gets added to the list of possible solutions.

Long story short, we've had too many `it` blocks. Those blocks are resource intensive according to Cypress's best practice documentation:  https://docs.cypress.io/guides/references/best-practices#Creating-tiny-tests-with-a-single-assertion. Our CI containers were running out of resources.

We used them as a way to group assertions and we loved seeing all those green checkmarks. The solution was to remove them where they were not absolutely needed. We converted most of them to `context` blocks, which don't show anything in stdout, but they help us during development to group assertions logically.

**Comment by stuart-clark** [comment]
In an effort to help anyone, here's a list of things folks have tried with some degree of success:

- Make sure you're using `timeout-minutes:` for your wallet and free tier
- Remove the use of `beforeEach` and/or `afterEach`
- Use [ec2-github-runner](https://github.com/machulav/ec2-github-runner)
- Run specs one at a time (basically specify `command:` in `cypress-io/github-action@v2` to call `for f in cypress/integration/*.test.js ; do cypress run --spec $f ; done`) per https://github.com/cypress-io/cypress/issues/8206#issuecomment-682721485
- `"video": false,` and/or`"videoUploadOnPasses": false,` per https://github.com/cypress-io/cypress/issues/17627#issuecomment-1059741645
- `"numTestsKeptInMemory": 0,` per https://github.com/cypress-io/cypress/issues/8206#issuecomment-745190829
- Limit `it` blocks per https://github.com/cypress-io/cypress/issues/8206#issuecomment-1086950214
- People are seeing various levels of success on older versions of Cypress, all the way back to `5.*.*`
- `DEBUG="cypress:*" ` https://github.com/cypress-io/cypress/issues/17627#issuecomment-1077619684
- Try different browsers
- Try docker
- Clear cookies cache and cookies in `beforeEach` per https://github.com/cypress-io/cypress/issues/8206#issuecomment-1419630366
- Add `cy.wait(1000)` or more at the end of the test to wait for all requests to finish per https://github.com/cypress-io/cypress/issues/8206#issuecomment-1685267659

I also saw some success using `timeout` (or `gtimeout`) while wrapping our Cypress command in [retry-action](https://github.com/marketplace/actions/retry-action) ... just mind parsing nested `:` characters unless you use @dmvict's latest fork.

🍻

_Before edit:_

> We are now configuring our runner to run on EC2 per [ec2-github-runner](https://github.com/machulav/ec2-github-runner) by @machulav. It requires a little AWS know-how, but it's a great resource.
> 
> **Using ec2-github-runner in addition to @kaiyoma's suggestion to run specs one file at a time allows our tests to pass quickly and reliably.**
>
> Being on EC2 lets you `ssh` into the instance, look around, troubleshoot, and control and allocate CPU and/or memory.
>
> However, even in a `t2.2xlarge` (32 GiB of memory and 8 vCPU) we still ran into timeouts unless we split our tests with @kaiyoma's method.
>
> My anecdotal feedback is that it seems highly correlated to `it` blocks and what happens-around-them, but I have never seen an exception nor did I see memory or CPU issues in our EC2 instance, even while debugging while Cypress was stalling. I noticed slowness in using `beforeEach` hooks, so we removed those too.

**Comment by Wandalen** [comment]
@stuart-clark you may use wretry.action[https://github.com/Wandalen/wretry.action] it's up to date

**Comment by martincarstens** [comment]
@stuart-clark Are you able to share your script for executing tests one at a time? We are currently doing this: https://gist.github.com/martincarstens/70dd91176e420c6c9ca3de96cafac71d - feel free to use it or adapt for your own use.

**Comment by viriatis** [comment]
I added this to my jenkins slave dockerfile and the problem was solved:

`# In summary, setting CI=true causes npm to set the Npm-In-CI header to true, 
# thus as a result the data gathered (by npm) assumes the package(s) are being installed via a "build farm", 
# (i.e. for Continuous Integration purposes), instead of a "human".
ENV CI=1 \
# disable shared memory X11 affecting Cypress v4 and Chrome
# https://github.com/cypress-io/cypress-docker-images/issues/270
  QT_X11_NO_MITSHM=1 \
  _X11_NO_MITSHM=1 \
  _MITSHM=0 \
  # point Cypress at the /root/cache no matter what user account is used
  # see https://on.cypress.io/caching
  CYPRESS_CACHE_FOLDER=/root/.cache/Cypress`

**Comment by viriatis** [comment]
I added this to my jenkins-slave dockerfile that I used to run cypress on jenkins and the build doens't get stuck anymore:

    `# In summary, setting CI=true causes npm to set the Npm-In-CI header to true, 
    # thus as a result the data gathered (by npm) assumes the package(s) are being installed via a "build farm", 
    # (i.e. for Continuous Integration purposes), instead of a "human".
    ENV CI=1 \
    # disable shared memory X11 affecting Cypress v4 and Chrome
    # https://github.com/cypress-io/cypress-docker-images/issues/270
      QT_X11_NO_MITSHM=1 \
      _X11_NO_MITSHM=1 \
      _MITSHM=0 \
      # point Cypress at the /root/cache no matter what user account is used
      # see https://on.cypress.io/caching
      CYPRESS_CACHE_FOLDER=/root/.cache/Cypress`

**Comment by mattvb91** [comment]
Upgraded to 9.6.0 and now started seeing this issue where all the tests succeed but cypress just hangs and never exits or sends a success signal. Just hangs there after displaying the test results.

**Comment by kutenai** [comment]
I'm seeing this also.. it's really causing us huge issues. Not sure what the issue is.. it needs to just FAIL if it won't work

**Comment by bporcelli** [comment]
We were experiencing this on Github Actions. Switching from Electron to Chrome resolved the hangs for us. Thanks @stuart-clark!

**Comment by liverant** [comment]
Any news? It's almost 2 years now...
This [issue](https://github.com/cypress-io/cypress/issues/17627 ) is on the same subject but a year after this one.

**Comment by mattvb91** [comment]
> We were experiencing this on GitHub Actions. Switching from Electron to Chrome resolved the hangs for us. Thanks @stuart-clark!

Thanks @bporcelli this fixed the issue for me.

**Comment by flotwig** [comment]
Users typically see this when the browser launched by Cypress uses too much memory and is automatically killed by the OS. I recommend folks who are impacted by this to follow #6170 and the related issue #349 which will greatly improve visibility of this issue.

**Comment by Komojo7** [comment]
Was experiencing this issue and switching to Firefox was the only thing that worked.

**Comment by cosmith** [comment]
I found this option that I thought would maybe help but it made no difference:
```
    /**
     * The number of tests for which snapshots and command data are kept in memory. Reduce this number if you are experiencing high memory consumption in your browser during a test run.
     * @default 50
     */
    numTestsKeptInMemory: number
```

> @stuart-clark Are you able to share your script for executing tests one at a time? We are currently doing this: https://gist.github.com/martincarstens/70dd91176e420c6c9ca3de96cafac71d - feel free to use it or adapt for your own use.

In the end this is the only thing that worked for us on CircleCI. We adapted the script to pass spec filenames split by CircleCI to parallelize runs.
It's 1.5x slower but at least it finishes successfully...

**Comment by stuart-clark** [comment]
This may be relevant to some folks here:

https://github.com/cypress-io/cypress/issues/17627#issuecomment-1199285216
https://github.com/cypress-io/cypress/pull/23001

**Comment by mattvb91** [comment]
Still the same on 10.10.0

```
The Test Runner unexpectedly exited via a exit event with signal SIGILL

Please search Cypress documentation for possible solutions:

https://on.cypress.io/

Check if there is a GitHub issue describing this crash:

https://github.com/cypress-io/cypress/issues

Consider opening a new issue.

----------

Platform: linux-x64 (Debian - 11.4)
Cypress Version: 10.10.0
```

**Comment by cosmith** [comment]
Update after a while, it turns out for us there was no issue apart from tests running for too long that were killed by CircleCI. So check that your retries param is not set too high and fix your flaky tests!

**Comment by mattvb91** [comment]
> Update after a while, it turns out for us there was no issue apart from tests running for too long that were killed by CircleCI. So check that your retries param is not set too high and fix your flaky tests!

in my case I cant even get the tests to start, this crashes before it even gets to the first test. The only stable cypress version so far is 9.7.0

**Comment by Neoxrus86** [comment]
Any news? I have this bug reproduced on versions 10.8.0, 12.2.0 and 12.3.0.
I don't know how to deal with it anymore.

**Comment by Neoxrus86** [comment]
Any news? I have this error reproduced in versions 10.8.0, 12.2.0 and 12.3.0.
I don't know how to deal with it anymore.
I'm running tests in Docker containers via gitlab-runner and the containers freeze up intermittently.

**Comment by GastonMujica** [comment]
Hey! To anyone still having this issue, this might help you out! 
We've tried almost everything, from reducing the amount of tests in memory, to the` --ipc=host` to` disable-dev-shm-usage`. Nothing worked. At the end we where able to **fix this issue by cleaning both the cache and the cookies.** 

```
export const clearCache = () => {
  cy.wrap(
    Cypress.automation('remote:debugger:protocol', {
      command: 'Network.clearBrowserCache',
    }),
  );
};

export const clearCookies = () => {
  cy.clearCookies();
};
```

Hope it helps!

**Comment by bahunov** [comment]
Is this done beforeEach I assume?

**Comment by GastonMujica** [comment]
> Is this done beforeEach I assume?

Yep, before anything else!

**Comment by Neoxrus86** [comment]
Hi all! I had a freeze inside a docker container. This solution helped me. I disable loading all images with intercept
```
    if (Cypress.env('disableMedia')) {
        cy.intercept(/\.(jpg|jpeg|png|mp4|gif|webp)+$/gi, []).as(
            'trash_requests_media'
        );
    }
```

**Comment by tausif29** [comment]
> Hey! To anyone still having this issue, this might help you out! We've tried almost everything, from reducing the amount of tests in memory, to the` --ipc=host` to` disable-dev-shm-usage`. Nothing worked. At the end we where able to **fix this issue by cleaning both the cache and the cookies.**
> 
> ```
> export const clearCache = () => {
>   cy.wrap(
>     Cypress.automation('remote:debugger:protocol', {
>       command: 'Network.clearBrowserCache',
>     }),
>   );
> };
> 
> export const clearCookies = () => {
>   cy.clearCookies();
> };
> ```
> 
> Hope it helps!



> Hey! To anyone still having this issue, this might help you out! We've tried almost everything, from reducing the amount of tests in memory, to the` --ipc=host` to` disable-dev-shm-usage`. Nothing worked. At the end we where able to **fix this issue by cleaning both the cache and the cookies.**
> 
> ```
> export const clearCache = () => {
>   cy.wrap(
>     Cypress.automation('remote:debugger:protocol', {
>       command: 'Network.clearBrowserCache',
>     }),
>   );
> };
> 
> export const clearCookies = () => {
>   cy.clearCookies();
> };
> ```
> 
> Hope it helps!

Hello @GastonMujica  I believe this code should be placed in `command.js`  , Could you please share where and how  to call these methods?

**Comment by vallme2003** [comment]
Confirming that this still happens in jenkins. Tried most of above solutions but doesn't work.

**Comment by FranciscoKnebel** [comment]
This should be a priority issue, just searching for Cypress and hanging/frozen returns years of responses with no actual solution.
I have tried every suggestion on this thread, and none of them worked for my case. Job runs with GitHub Actions in a self-hosted runner, with 8 GB of RAM, which should be more than enough.

A temporary solution I made to reduce the effects was to create multiple separate jobs using a [matrix of jobs](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs), testing each spec in different jobs. Still has the frozen processes, but at least it doesn't hang the entire test suite.

**Comment by amakhrov** [comment]
It is affecting us badly lately. It used to be a rare case, but seems to have become worse for the last two months or so.

The debug output (via `DEBUG='cypress:*'`) doesn't show anything suspicious. It's just that at some point the tests stop executing at all.
The browser seems to be running for some time after that (there is a periodic network request issued by the last opened page). After some time it stops, too - and the only output produced is CPU/memory profiler.

Looks like this:

```
2023-06-29T00:04:26.131Z cypress:server:remote-states getting remote state: { auth: null, origin: '<EDIT>', strategy: 'http', fileServer: null, domainName: '<EDIT>', props: { port: '443', protocol: 'https:', subdomain: '<EDIT>', domain: '<EDIT>', tld: 'io' } } for: <EDIT>
2023-06-29T00:04:27.758Z cypress:server:video capture codec data: { format: 'image2pipe', audio: '', video: 'mjpeg (Baseline)', duration: 'N/A', video_details: [ 'mjpeg (Baseline)', 'yuvj420p(pc', 'bt470bg/unknown/unknown)', '1280x720 [SAR 1:1 DAR 16:9]', '25 fps', '25 tbr', '25 tbn', '25 tbc' ] }
2023-06-29T00:04:28.821Z cypress:server:util:process_profiler current & mean memory and CPU usage by process group:
┌─────────┬───────────────────┬──────────────┬──────────────────────────┬────────────┬────────────────┬──────────┬──────────────┬─────────────┐
│ (index) │       group       │ processCount │           pids           │ cpuPercent │ meanCpuPercent │ memRssMb │ meanMemRssMb │ maxMemRssMb │
├─────────┼───────────────────┼──────────────┼──────────────────────────┼────────────┼────────────────┼──────────┼──────────────┼─────────────┤
│    0    │     'cypress'     │      1       │          '5842'          │    3.55    │      4.33      │  444.29  │    430.59    │   467.24    │
│    1    │    'Electron'     │      1       │          '7305'          │     2      │      3.22      │  399.79  │    413.27    │    522.8    │
│    2    │ 'electron-shared' │      4       │ '5878, 6061, 5879, 6137' │    5.24    │      2.84      │  266.69  │    248.81    │   276.31    │
│    3    │     'plugin'      │      1       │          '6127'          │     0      │      0.03      │  247.06  │    228.53    │   385.09    │
│    4    │     'ffmpeg'      │      1       │          '7319'          │    3.26    │      2.01      │  139.73  │    117.53    │   140.15    │
│    5    │      'other'      │      2       │       '7495, 7496'       │     0      │       0        │   3.86   │     3.74     │    3.86     │
│    6    │      'TOTAL'      │      10      │           '-'            │   14.05    │     12.24      │ 1501.43  │   1399.71    │   1623.33   │
└─────────┴───────────────────┴──────────────┴──────────────────────────┴────────────┴────────────────┴──────────┴──────────────┴─────────────┘
2023-06-29T00:04:29.796Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 56684 }
2023-06-29T00:04:29.807Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 42114 }
2023-06-29T00:04:29.807Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46274 }
2023-06-29T00:04:29.825Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46292 }
2023-06-29T00:04:29.946Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46320 }
2023-06-29T00:04:29.946Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46306 }
2023-06-29T00:04:29.949Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46330 }
2023-06-29T00:04:30.062Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46352 }
2023-06-29T00:04:30.194Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46254 }
2023-06-29T00:04:30.340Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46278 }
2023-06-29T00:04:30.462Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46258 }
2023-06-29T00:04:30.508Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46342 }
2023-06-29T00:04:31.138Z cypress:server:util:socket_allowed allowed socket closed, removing { localPort: 46356 }
2023-06-29T00:04:39.079Z cypress:server:util:process_profiler current & mean memory and CPU usage by process group:
┌─────────┬───────────────────┬──────────────┬──────────────────────────┬────────────┬────────────────┬──────────┬──────────────┬─────────────┐
│ (index) │       group       │ processCount │           pids           │ cpuPercent │ meanCpuPercent │ memRssMb │ meanMemRssMb │ maxMemRssMb │
├─────────┼───────────────────┼──────────────┼──────────────────────────┼────────────┼────────────────┼──────────┼──────────────┼─────────────┤
│    0    │     'cypress'     │      1       │          '5842'          │    2.18    │      4.18      │  415.82  │    429.54    │   467.24    │
│    1    │    'Electron'     │      1       │          '7305'          │    1.25    │      3.06      │  380.07  │    410.5     │    522.8    │
│    2    │ 'electron-shared' │      4       │ '5878, 6061, 5879, 6137' │    5.71    │      3.05      │  264.29  │    249.92    │   276.31    │
│    3    │     'plugin'      │      1       │          '6127'          │    0.09    │      0.03      │  171.81  │    224.16    │   385.09    │
│    4    │     'ffmpeg'      │      1       │          '7319'          │    1.32    │      1.96      │  139.98  │    119.26    │   140.15    │
│    5    │      'other'      │      2       │       '7579, 7580'       │     0      │       0        │   3.71   │     3.73     │    3.86     │
│    6    │      'TOTAL'      │      10      │           '-'            │   10.54    │     12.12      │ 1375.68  │   1397.99    │   1623.33   │
└─────────┴───────────────────┴──────────────┴──────────────────────────┴────────────┴────────────────┴──────────┴──────────────┴─────────────┘
[6061:0629/000445.710042:ERROR:gl_utils.cc(319)] [.WebGL-0x3a4800345500]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, High): GPU stall due to ReadPixels
2023-06-29T00:04:47.223Z cypress:server:server-base Got CONNECT request from collector-pxz6penin0.px-cloud.net:443
2023-06-29T00:04:47.223Z cypress:https-proxy Writing browserSocket connection headers { url: 'collector-pxz6penin0.px-cloud.net:443', headLength: 0, headers: { host: 'collector-pxz6penin0.px-cloud.net:443', 'proxy-connection': 'keep-alive', 'user-agent': 'Mozilla/5.0 (X11; Linux x86_64) 
```

The last messages above are from that periodic network request (from a PerimeterX sensor). Nothing test-specific is printed after that, just Perimeterx requests interleaved with CPU profiling

Cypress 12.11 with Electron

**Comment by martincarstens** [comment]
🔔 Reposting this for visibility as it might get have gotten buried under the other comments. We use this script to run our Cypress tests: https://gist.github.com/martincarstens/70dd91176e420c6c9ca3de96cafac71d

It's a simple script that auto detects hangs and prevents runaway jobs. Since implementing it over a year ago, Cypress has been stable for us. We have 157 spec files running in various configurations in GitHub Actions with parallel jobs. We are using the latest version of Cypress.

To be fair, we shouldn't need to be using this script, but it is the only thing that returned sanity to our workflows.

@stuart-clark might be able to provide some updates running Cypress with the script in Circle CI.

@cosmith any updates from your side?

Rooting for the Cypress team to put this one to bed.

**Comment by cosmith** [comment]
Hello, on our side we migrated to Playwright. 

I do remember that in the end the "no output" that we were seeing in CircleCI was actually a "very long output" when some tests were failing and retrying several times in a row. So it was not really a bug in Cypress (even though it would have been a lot easier to debug if there was some log stating that the test was restarting).

**Comment by martincarstens** [comment]
> Hello, on our side we migrated to Playwright.
> 
> I do remember that in the end the "no output" that we were seeing in CircleCI was actually a "very long output" when some tests were failing and retrying several times in a row. So it was not really a bug in Cypress (even though it would have been a lot easier to debug if there was some log stating that the test was restarting).

We've also started migrating to Playwright, but it's taking us a bit because of all the effort invested into writing Cypress specs. Thanks for the update.

**Comment by FranciscoKnebel** [comment]
This bug is very breaking for my use case, so I'm studying a soft migration to Playwright now as well, at least the non-working tests for now.

We also have a big problem caused by how Cypress handles cross-origin and iframe tests, which unfortunately is not stable enough right now, but some test cases are very dependant on it.

This is a real bummer, since the last couple of months I spent on actually migrating our tests from another platform to Cypress, so finding out some very crucial steps are impossible is really demotivating.

For the others watching the thread, some links I read on this migration:

[Migration from Cypress to Playwright - hype or great?](https://www.21risk.com/blog/migration-from-cypress-to-playwright-hype-or-great) 

[On Migrating from Cypress to Playwright](https://mtlynch.io/notes/cypress-vs-playwright/) 

[Migrating from Cypress to Playwright](https://dev.to/lucgagan/migrating-from-cypress-to-playwright-pn0) 

[Rewriting tests from Cypress to Playwright using GPT3 by Gajus Kuizinas](https://contra.com/p/PWBcPYZc-rewriting-tests-from-cypress-to-playwright-using-gpt-3) 

[Convert Cypress to Playwright](https://ray.run/tools/cypress-to-playwright)  (tool built for this migration) 

Hopefully the Cypress team fixes these crucial things eventually, but reading over a year old (in this case, almost three) posts reporting the same problem to something as important as just running tests, in a test framework, and that functionality not working, is a deal-breaker.

**Comment by SIGSTACKFAULT** [comment]
also affecting me in GitLab CI. Using the `cypress/browsers` Docker image.
```
Cypress package version: 12.17.0
Cypress binary version: 12.17.0
Electron version: 21.0.0
Bundled Node version: 16.16.0
```

**Comment by alexszilagyi** [comment]
Any updates on this?

**Comment by amakhrov** [comment]
We ended up patching the runner (we currently use `cypress-cloud`) to implement timeout/retry per spec. Hacky af, but our CI pipelines are much more stable now with this change.

**Comment by GarrisonD** [comment]
**TLDR:** Add `cy.wait(1000)` or more at the end of the test to wait for all requests to finish.

I have an observation that could assist the Cypress team in fixing a potential bug in Cypress (if confirmed as a bug), or individuals like myself in applying a temporary workaround to address the issue.

I came across this thread when my test suite began encountering issues during execution in the CI environment, sometimes stalling midway and necessitating a restart. While restarting occasionally resolved the problem, it wasn't a consistent solution. Despite my attempts at investigation, I was unable to pinpoint the root cause.

Subsequently, I upgraded Cypress to the latest version and encountered a similar issue with a very straightforward test involving the `Sign In` form. This test includes visiting `localhost`, inputting login credentials, submitting the form, and then verifying the redirected URL.

The `sign-in.cy.ts` file wouldn't complete execution in Cypress. It failed to display the statistics for the `passed`/`failed`/`skipped` tests; the process simply hung. Given the simplicity of the test, it became evident that the problem lay with Cypress itself.

I speculated that the issue might be related to requests spawned by the main view of the application, to which users are redirected after a successful sign-in. As a quick workaround, I appended `cy.wait(1000)` to the test's end, and to my delight, it resolved the problem 🎉 

However, I tried to find a more elegant solution utilizing `cy.intercept('/api/**').as('api')` followed by `cy.wait('@api')`. Regrettably, this approach didn't yield the desired outcome.

Consequently, I am currently relying on the slightly inelegant but functional `cy.wait(1000)` workaround, with the hope that this issue will eventually be addressed within Cypress itself. 💡 It's possible that all pending requests could be automatically canceled at the end of each test...

**Comment by bahunov** [comment]
Thanks for this valuable input!!!

**Comment by stuart-clark** [comment]
> @stuart-clark might be able to provide some updates running Cypress with the script in Circle CI.

Hi Martin! 👋 To clarify, we've always used Github Actions, not Circle CI.

We don't run into this issue anymore. So we are stable, some notes on our configuration off the top of my head:
- We are using Cypress 12 (I think this is important for [**Test Isolation**](https://docs.cypress.io/guides/core-concepts/test-isolation)) https://www.cypress.io/blog/2022/12/06/announcing-cypress-12/
- We are not using `ec2-github-runner` anymore, we run on Github Actions with "larger runners" https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners#running-jobs-on-your-runner. I had started to see the same behavior with failures on the ec2 runners.
- We use matrixed runners; we are using Currents for orchestration
- Entire suite finishes in about 5 minutes ✨

While I still think there are issues here, we get around them by being matrixed in such a way that a runner is only running 3-6 spec files of very few `it` blocks each. Our largest spec file only has 6 `it` blocks, and each runner only does about 5-6 tests. 

To do this efficiently (💸) on Github Actions one needs to leverage `actions/cache` (https://github.com/actions/cache) and other strategies so you're not spending a ton of time on setup for a large amount of separate runners.

For anyone facing this, I would really just suggest **breaking your tests up as much as possible** and using a matrix strategy or a script similar to what @martincarstens wrote.

**Comment by Phonesis** [comment]
Thanks @GarrisonD 

We discovered this today as part of using the new Test Replay feature with Cypress v13. There was always a 1 min wait in our after hook after an assertion around network responses failed. 

It is definitely far from ideal adding a wait in like this. There is clearly a bug here around disposing of network call promises or something perhaps

**Comment by bahunov** [comment]
@GarrisonD <https://github.com/GarrisonD>
You mean add afterEach?

**Comment by GarrisonD** [comment]
@bahunov I have more than one test in a file, so yeah, I use `afterEach` hook.

The key point here is to wait for running requests, where and how is a minor thing.

**Comment by zeldigas** [comment]
Just want to share one more workaround for hang problem. In our case we are running files individually, and hang occured after test completion - for some reason cypress failed to properly exit. 

I ended up with implementing parallel watchdog with bash to track this.

Gist: https://gist.github.com/zeldigas/6d2041c51b96b98b4859017f55a9bc34

**Comment by jennifer-shehane** [comment]
@Phonesis Please make sure to update to the latest version of the Cypress App. We've been actively addressing hanging issues related to Test Replay.

**Comment by jennifer-shehane** [comment]
Hi everyone, stalling and hanging within Cypress could be attributed to many different reasons. We do work through hangs/stalls issues as quickly as we can since we know this is the worst situation that your tests can be stuck in and it is an extra expense for CI. The causes we've found for hanging have been very wide and we do need clear reproducible examples provided in order to investigate.

I suggest opening a new issue describing any current hanging issues and as much info as you can provide for us to investigate, including debug logs. Everyone's issue in this thread could be caused by very different things and unrelated. I'll be closing this issue since it is not encompassing a single cause that we can work towards a resolution on and close. We have fixed several causes of hanging issues over the years that might have addressed many of these as well.

**Comment by amakhrov** [comment]
> Everyone's issue in this thread could be caused by very different things and unrelated

How would one know the root cause for their specific case?
The symptoms in all cases are the same: the thing just gets stuck with no errors / other indication of smth being wrong.

**Comment by madardour** [comment]
I have also this issue.
I have 40 test (feature) files with parallel processing using the plugin cypress-parallel. For long time i have running cypress e2e tests in AzureDevops on version 9.x without issues. Begin this year i have upgraded to v12.4. Upgrade was necessary for me because of the proxy. v9.x (or i ) was not able to call webserver behind a proxy.
Since upgrade to v12 the cypress process hangs. The process still appears in TaskManager but it does nothing and AzureDevops waits for until i kill the proces manually.
i have tried several versions (12.4, 12.10, 12.13, 12.15, 13,0) and the issue still occurs.

**Comment by nagash77** [comment]
@madardour please open a new issue with a [reproducible example](https://github.com/cypress-io/cypress-test-tiny) and the Cypress team will be happy to investigate.

**Comment by elenasch** [comment]
I have been running into a similar issue with Cypress just hanging. In my case it was on a very particular new test suite that was added. What we found is that the app UI had performance issues with excessive DOM element counts. 
This perf issue didn't surface to naked eye, but crashed several of our QA automation systems and frameworks (Jest, Selenium and Cypress). The Cypress side effect was a handing runner. 

Hope that helps with reproduction and fixing this scenario.

Cypress `12.7.0`
It didn't matter how large or small machine resources running these tests were. The runner hung the same way, and I could see the memory of chrome browser growing over time, but other processes didn't get affected as much. 

We have ~4500 DOM elements on initial render and that number kept growing  (anything over ~1500 is known to be a possible perf bottleneck)
